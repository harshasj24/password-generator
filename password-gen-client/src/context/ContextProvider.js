import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";

export const passwordCotext = createContext();

export const ContextProvider = ({ children }) => {
  const [sliderVal, setSliderVal] = useState(12);
  const [password, setPassword] = useState("");
  const [breakPoint, setBreakPoint] = useState(null);
  const [isLoading, setisloading] = useState(false);
  const breakPoints = {
    mobile: "(max-width:600px)",
    tablet: "(min-width:600px) and (max-width:960px)",
    laptop: "(min-width:960px) and (max-width:1023px)",
    desktop: "(min-width:1024px)",
  };

  const copyToClipBoard = (text) => {
    window.navigator.clipboard.writeText(text);
  };

  const matchMediaQuerry = (breakPoints, setBreakPoint) => {
    Object.keys(breakPoints).map((key) => {
      if (window.matchMedia(breakPoints[key]).matches) {
        setBreakPoint(key);
      }
    });
  };

  const breakPointObserver = () => {
    window.addEventListener("resize", () => {
      matchMediaQuerry(breakPoints, setBreakPoint);
    });
  };

  const values = useMemo(() => {
    return {
      password,
      setPassword,
      sliderVal,
      setSliderVal,
      copyToClipBoard,
      breakPointObserver,
      breakPoint,
      isLoading,
      setisloading,
    };
  });

  return (
    <div>
      <passwordCotext.Provider value={values}>
        {children}
      </passwordCotext.Provider>
    </div>
  );
};

export const usePassword = () => {
  return useContext(passwordCotext);
};
