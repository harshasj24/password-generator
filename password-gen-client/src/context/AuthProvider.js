import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useNavigate } from "react-router-dom";
import { useLocalStorage } from "../hooks/UseLocalStorage";
import axios from "axios";
import { useVault } from "./VaultProvider";
const AuthContext = createContext();

export const AuthProvider = ({ children }) => {
  // const navigate = useNavigate();
  // const [isLoggedIn, setIsLoggedin] = useState(false);
  const [user, setUser] = useLocalStorage("user", null);
  useEffect(() => {
    console.log(user);
  }, [user]);
  const login = async (userDetails) => {
    setUser(userDetails);
    console.log(userDetails, "auth-provider");
  };
  const logout = async () => {
    setUser(null);
    // console.log(responce);
  };
  const value = useMemo(() => {
    return { user, login, logout, setUser };
  }, [user, login]);
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export const useAuth = () => {
  return useContext(AuthContext);
};
