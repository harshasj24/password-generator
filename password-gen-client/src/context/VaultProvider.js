import axios from "axios";
import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useNavigate } from "react-router-dom";
import { useAuth } from "../context/AuthProvider";
import environemnt from "../environments/environment";
const vaultContex = createContext();
export const VaultProvider = ({ children }) => {
  const [passwords, setPasswords] = useState([]);
  const { user } = useAuth();
  const navigate = useNavigate();
  const baseUrl = environemnt.apiUrl;
  const [updatePass, setUpdatePass] = useState({
    action: null,
    isUpdating: false,
    _id: null,
    pName: "",
  });
  const [snackbar, setSnackbar] = useState({
    open: false,
    message: "",
  });
  const [modalDetails, setModalDetails] = useState({
    open: false,
    _id: null,
    type: "",
  });

  const handleModalDetails = (details) => {
    setModalDetails({ ...modalDetails, ...details });
  };

  const handleSnackbar = (open, message) => {
    setSnackbar({ ...snackbar, open: open, message: message });
  };
  const [slectedId, setSelectedId] = useState(null);
  useEffect(() => {
    console.log(user, "effect");
  }, [user]);

  const headers = user && {
    Authorization: "barrer " + user.token,
  };

  const savePassword = async (PasswordDetails) => {
    try {
      const repsonce = await axios.post(
        `${baseUrl}/vault/savePasswords`,
        { ...PasswordDetails, _id: user.id },
        {
          headers: headers,
        }
      );
      const newPassword = {
        ...PasswordDetails,
        _id: repsonce.data.data._id,
        date: repsonce.data.data.date,
        password: repsonce.data.data.cipherText,
        status: repsonce.data.data.status,
      };

      setPasswords([...passwords, newPassword]);
    } catch (error) {
      console.log(error);
    }
  };
  const isUpdating = (details) => {
    setUpdatePass({
      ...updatePass,
      ...details,
    });
  };
  const resetUpdating = () => {
    setUpdatePass({
      ...updatePass,
      isUpdating: false,
      _id: null,
      pName: "",
    });
  };
  const updatePassword = async (passwordDetails) => {
    const { password, pName } = passwordDetails;
    const reqData = {
      password,
      _id: updatePass._id,
      pName,
    };
    const responece = await axios.put(
      `${baseUrl}/vault/updatePassword`,
      reqData,
      {
        headers: headers,
      }
    );

    const index = passwords.findIndex((data) => {
      return data._id === updatePass._id;
    });
    console.log(index);

    const newPasswords = [...passwords];
    newPasswords[index] = {
      ...newPasswords[index],
      password: responece.data.data.cipherText,
      date: responece.data.data.date,
      status: responece.data.data.status,
      pName,
    };
    setPasswords([...newPasswords]);
    handleSnackbar(true, "Password Updated Sucessfully");
  };

  const getAllPasswords = async () => {
    try {
      if (passwords.length === 0) {
        const repsonce = await axios.get(
          `${baseUrl}/vault/password/${user?.id}`,
          {
            headers: headers,
          }
        );
        console.log(repsonce);

        setPasswords(repsonce.data.data);
      }
    } catch (error) {
      console.log(error);
      setPasswords([]);
    }
  };
  // validatePassword
  const checkPassword = async (data) => {
    try {
      const responce = await axios.post(
        `${baseUrl}/users/validatePassword`,
        data,
        {
          headers: headers,
        }
      );
      return responce;
    } catch (error) {
      if (error.status === 403) {
        navigate("/access");
      }
      return false;
    }
  };

  const deletePassword = async () => {
    try {
      const responce = await axios.delete(
        `${baseUrl}/vault/delete/${updatePass._id}`,
        {
          headers: headers,
        }
      );
      const newPasswords = [...passwords].filter(
        (val) => val._id !== updatePass._id
      );
      setPasswords(newPasswords);
      handleSnackbar(true, "Password Deleted Sucessfully");
    } catch (error) {
      console.log(error);
    }
  };

  const loopElements = (range, element) => {
    for (let i = 0; i < range; i++) {
      console.log(i, range, "vault");
      return element;
    }
  };

  const getOnePassword = async (_id) => {
    const responce = await axios.get(`${baseUrl}/vault/getPassword/${_id}`, {
      headers: headers,
    });
    return responce;
  };
  const values = useMemo(() => {
    return {
      passwords,
      setPasswords,
      savePassword,
      getAllPasswords,
      updatePass,
      isUpdating,
      updatePassword,
      headers,
      resetUpdating,
      checkPassword,
      slectedId,
      setSelectedId,
      deletePassword,
      pName: updatePass.pName,
      loopElements,
      getOnePassword,
      snackbar,
      handleSnackbar,
      modalDetails,
      handleModalDetails,
    };
  }, [
    passwords,
    updatePass,
    user,
    slectedId,
    snackbar.open,
    modalDetails,
    handleSnackbar,
  ]);

  return <vaultContex.Provider value={values}>{children}</vaultContex.Provider>;
};

export const useVault = () => {
  return useContext(vaultContex);
};
