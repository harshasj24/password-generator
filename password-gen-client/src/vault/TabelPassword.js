import {
  Button,
  CircularProgress,
  IconButton,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from "@mui/material";
import React, { useEffect, useState } from "react";
import LaunchIcon from "@mui/icons-material/Launch";

import { formatDate, formatTime } from "./customdate";
import { usePassword } from "../context/ContextProvider";
import { Link, useNavigate } from "react-router-dom";
import Launch from "@mui/icons-material/Launch";
import { useVault } from "../context/VaultProvider";
import axios from "axios";
import { useAuth } from "../context/AuthProvider";
import EditIcon from "@mui/icons-material/Edit";
import DeleteIcon from "@mui/icons-material/Delete";
import EditOutlinedIcon from "@mui/icons-material/EditOutlined";
import DeleteOutlineIcon from "@mui/icons-material/DeleteOutline";
import CircularProgressWithLabel from "./CircularProgressWithLabel";
import { Facebook, NavigateBefore, Twitter } from "@mui/icons-material";
import { links, checkLinks } from "./constant";
import MoreHorizIcon from "@mui/icons-material/MoreHoriz";
const TabelPassword = ({ passwords, setModal, search, handleModalOpen }) => {
  const { copyToClipBoard } = usePassword();
  const { updatePass, isUpdating, headers, setSelectedId, getOnePassword } =
    useVault();

  const navigate = useNavigate();
  const [currentDate, setCurrentDate] = useState(new Date());
  const getPercentage = (value) => {
    return (value / 50) * 100;
  };

  const handelClick = (_id) => async (e) => {
    // console.log(_id);
    // e.target.innerText = "....";
    // const responce = await axios.get(
    //   `http://localhost:4500/vault/getPassword/${_id}`,
    //   {
    //     headers: headers,
    //   }
    // );
    // copyToClipBoard(responce.data.data.password);
    // e.target.innerText = "Copied";
    // setTimeout(() => {
    //   e.target.innerText = "copy";
    // }, 2000);
    // // <CircularProgress size="23px" />
  };
  useEffect(() => {
    let interval = setInterval(() => {
      setCurrentDate(new Date());
    }, 60000);
    return () => {
      clearInterval(interval);
    };
  }, []);
  const modifieDate = (date) => {
    const newDate = new Date(date);

    var diff = Math.trunc(
      (currentDate.getTime() - newDate.getTime()) / 1000 / 60
    );

    if (diff < 1) {
      return "Just now";
    } else if (diff > 0 && diff <= 60) {
      return diff + "min ago";
    } else if (diff > 60 && diff < 1440) {
      return Math.trunc(diff / 60) + "hr ago";
    } else if (diff >= 1440) {
      return `${formatDate(newDate)}\n${formatTime(newDate)}`;
    }
  };

  const handleLinkClick = (_id, pName) => () => {
    isUpdating({ _id: _id, action: "edit", pName: pName });
    navigate("/access");
  };
  const getColor = (status) => {
    console.log(status);
    return status >= 12
      ? "success"
      : status > 5 && status <= 10
      ? "warning"
      : "error";
  };

  const handleDelete = (_id) => () => {
    isUpdating({ _id: _id, action: "delete" });
    navigate("/access");
  };
  return (
    <div>
      <TableContainer sx={{ height: "410px" }} component={Paper}>
        <Table stickyHeader>
          <TableHead sx={{ color: "white" }}>
            <TableRow>
              <TableCell>
                {" "}
                <h6>Website</h6>{" "}
              </TableCell>
              <TableCell>
                <h6> password</h6>{" "}
              </TableCell>
              <TableCell>
                {" "}
                <h6>Last Modified</h6>{" "}
              </TableCell>
              <TableCell>
                {" "}
                <h6>Status</h6>{" "}
              </TableCell>
              <TableCell>
                <h6>Actions</h6>{" "}
              </TableCell>
            </TableRow>
          </TableHead>
          <TableBody data-testid="tableBody">
            {passwords.filter((data) =>
              data.pName.toLowerCase().includes(search.toLowerCase())
            ).length > 0 ? (
              passwords
                .filter((data) =>
                  data.pName.toLowerCase().includes(search.toLowerCase())
                )
                .map((data, i) => (
                  <TableRow key={i}>
                    <TableCell>
                      {data.pName}
                      {checkLinks(data.pName) && (
                        <a
                          className="ml-2"
                          target="_blank"
                          href={checkLinks(data.pName)?.link}
                        >
                          <LaunchIcon fontSize="inherit" />
                        </a>
                      )}
                    </TableCell>
                    <TableCell>
                      <Button
                        onClick={handleModalOpen(data._id)}
                        data-testid={`copy${i}`}
                        size="small"
                        variant="outlined"
                      >
                        view/copy
                      </Button>
                      {/* <IconButton
                      onClick={handleModalOpen(data._id)}
                      data-testid={`copy${i}`}
                      size="small"
                      variant="outlined"
                    >
                      <MoreHorizIcon />
                    </IconButton> */}
                    </TableCell>
                    <TableCell>{modifieDate(data.date)}</TableCell>
                    <TableCell>
                      <CircularProgressWithLabel
                        color={getColor(data.status)}
                        value={getPercentage(data.status)}
                      />
                    </TableCell>
                    <TableCell>
                      <IconButton
                        size="small"
                        onClick={handleLinkClick(data._id, data.pName)}
                      >
                        <EditOutlinedIcon color="primary" fontSize="small" />
                      </IconButton>
                      <IconButton
                        onClick={handleDelete(data._id)}
                        sx={{ ml: 2 }}
                        size="small"
                      >
                        <DeleteOutlineIcon color="error" fontSize="small" />
                      </IconButton>
                      {/* <Button onClick={handleLinkClick(data._id)}>
                    change <Launch fontSize="inherit" />
                  </Button> */}
                    </TableCell>
                  </TableRow>
                ))
            ) : (
              <TableRow>
                <TableCell colSpan={5} className="text-center">
                  No Data Found
                </TableCell>
              </TableRow>
            )}
          </TableBody>
        </Table>
      </TableContainer>
    </div>
  );
};

export default TabelPassword;
