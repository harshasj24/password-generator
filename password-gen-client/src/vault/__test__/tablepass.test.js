import { fireEvent, render, screen } from "@testing-library/react";
import { AuthProvider } from "../../context/AuthProvider";
import { VaultProvider } from "../../context/VaultProvider";
import TabelPassword from "../TabelPassword";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import { ContextProvider } from "../../context/ContextProvider";
import { BrowserRouter } from "react-router-dom";
import MockAapter from "axios-mock-adapter";
import axios from "axios";
let component;
const passwords = [
  {
    pName: "G pay",
    _id: "1234ytfhsd567",
    date: new Date(),
  },
  {
    pName: "paytm",
    _id: "1234ytfhsd56785t",
    date: new Date(),
  },
  {
    pName: "phone pay",
    _id: "1234ytfhsd569w",
    date: new Date(),
  },
];
const setOpen = jest.fn();
beforeEach(() => {
  component = (
    <BrowserRouter>
      <AuthProvider>
        <ContextProvider>
          <VaultProvider>
            <TabelPassword passwords={passwords} setModal={setOpen} />
          </VaultProvider>
        </ContextProvider>
      </AuthProvider>
    </BrowserRouter>
  );
  render(component);
});

describe("Testing Table password component", () => {
  it("it should render the table password component", () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("it should contain the table rows", () => {
    const tableBody = screen.getByTestId("tableBody");
    expect(tableBody).toBeTruthy();
    expect(tableBody.childNodes.length).toEqual(3);
  });
  it("it should call the copy button", () => {
    const btn = screen.getByTestId("copy0");
    const handleClick = jest.fn();
    fireEvent.click(btn, { Event: handleClick() });
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
  it("it should call the get API", () => {
    let mock = new MockAapter(axios);
    mock
      .onGet("http://localhost:4500/vault/getPassword/1234567")
      .reply(200, passwords);
  });
  axios
    .get("http://localhost:4500/vault/getPassword/${_id}")
    .then((responce) => {
      expect(responce.data).toBe("passwords");
    });
});
