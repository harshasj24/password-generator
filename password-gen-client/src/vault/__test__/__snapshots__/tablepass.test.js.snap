// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Table password component it should render the table password component 1`] = `
<div>
  <div>
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-v0tj71-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        className="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
        role={null}
      >
        <thead
          className="MuiTableHead-root css-197n1ah-MuiTableHead-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
            >
               
              <h6>
                Website
              </h6>
               
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
            >
              <h6>
                 password
              </h6>
               
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
            >
               
              <h6>
                Last Modified
              </h6>
               
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
            >
               
              <h6>
                Status
              </h6>
               
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
            >
              <h6>
                Actions
              </h6>
               
            </th>
          </tr>
        </thead>
        <tbody
          className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          data-testid="tableBody"
          role={null}
        >
          <tr
            className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium text-center css-1ex1afd-MuiTableCell-root"
              colSpan={5}
            >
              No Data Found
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
